File: Lab5/ACF.m
function acf=ACF(src, fs, framesize)
    sz = size(src, 1);  
    
    % Params
    n = framesize;
    startFrame = 1;
    frameNum = floor(sz/n);

    acf = zeros(frameNum, 1);

    for f=startFrame:frameNum-1
        start = (f-1)*n+1;
        Frame = src(start:start+n);
        tAcf = zeros(n, 1);
        for t=10:n
            ShiftFrame = src(start+t:start+n+t);
            tAcf(t) = dot(Frame(1:n-1-t), ShiftFrame(1:n-1-t));
        end
        if f==8
            figure;
            plot(tAcf(10:end));
            break;
        end
        [~,i] = max(tAcf);
        acf(f) = fs/i;
    end
end

File: Lab5/ACF.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jan 26 13:07:50 2017
#獨奏音樂之音高偵測(AMDF與Auto-correlation方法)。準備不同獨奏音樂一分鐘音檔兩個。
#ACF : Auto-Correlation Function
#use python3
@author: Lulu
"""
import numpy as np
import pylab as P
import math as m
import scipy.io.wavfile
from scipy.ndimage.interpolation import shift
FSample, samples = scipy.io.wavfile.read("01-1.wav")
samples = samples.astype(int)

length = (len(samples))

framesize = 1024
startFrame = 0
#frameNum = 1
frameNum = m.floor(length/framesize)

index = 0
Frame = np.zeros(framesize,dtype=np.int32)
ShiftFrame = np.zeros(framesize,dtype=np.int32)
acf = np.zeros(framesize,dtype=np.int32)
acfFinal = np.zeros(frameNum,dtype=np.int32)

i = 0
j = 0
index = 0

#autocorelation
for index in range(0,frameNum,1):
    Frame = samples[(startFrame+index)*framesize:(startFrame+index+1)*framesize]
    #Frame = samples[48000:48000+framesize]
    ShiftFrame = Frame
    acf = np.zeros(framesize,dtype=np.int32)
    for i in range(0,framesize,1):
        ShiftFrame = shift(Frame, i, cval=0)
        for j in range(i,framesize,1):
            acf[i] += Frame[j] * ShiftFrame[j]
    acf[0:5] = -10000000
    acfFinal[index] = np.argmax(acf)
            
        
print(acfFinal)
time = np.zeros(framesize,dtype=np.int16)
for a in range(0,framesize,1):
    time[a] = a

P.plot(time,acf)
P.xlabel('time (s)')
P.ylabel('acf')